src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs

```diff
@@ -0,0 +1,1864 @@
+        /// <returns>The value associated with the specified key. If the specified key is not found, a get operation throws a <see cref="KeyNotFoundException"/>, and a set operation creates a new element with the specified key.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="key"/> is null.</exception>
+        /// <exception cref="KeyNotFoundException">The property is retrieved and <paramref name="key"/> does not exist in the collection.</exception>
+        /// <remarks>Setting the value of an existing key does not impact its order in the collection.</remarks>
+        public TValue this[TKey key]
+        {
+            get
+            {
+                if (!TryGetValue(key, out TValue? value))
+                {
+                    ThrowHelper.ThrowKeyNotFound(key);
+                }
+
+                return value;
+            }
+            set
+            {
+                ArgumentNullException.ThrowIfNull(key);
+
+                bool modified = TryInsert(-1, key, value, InsertionBehavior.OverwriteExisting);
+                Debug.Assert(modified);
+            }
+        }
+
+        /// <summary>Insert the key/value pair at the specified index.</summary>
+        /// <param name="index">The index at which to insert the pair, or -1 to append.</param>
+        /// <param name="key">The key to insert.</param>
+        /// <param name="value">The value to insert.</param>
+        /// <param name="behavior">
+        /// The behavior controlling insertion behavior with respect to key duplication:
+        /// - IgnoreInsertion: Immediately ends the operation, returning false, if the key already exists, e.g. TryAdd(key, value)
+        /// - OverwriteExisting: If the key already exists, overwrites its value with the specified value, e.g. this[key] = value
+        /// - ThrowOnExisting: If the key already exists, throws an exception, e.g. Add(key, value)
+        /// </param>
+        /// <returns>true if the collection was updated; otherwise, false.</returns>
+        private bool TryInsert(int index, TKey key, TValue value, InsertionBehavior behavior)
+        {
+            // Search for the key in the dictionary.
+            uint hashCode = 0, collisionCount = 0;
+            int i = IndexOf(key, ref hashCode, ref collisionCount);
+
+            // Handle the case where the key already exists, based on the requested behavior.
+            if (i >= 0)
+            {
+                Debug.Assert(_entries is not null);
+
+                switch (behavior)
+                {
+                    case InsertionBehavior.OverwriteExisting:
+                        _entries[i].Value = value;
+                        return true;
+
+                    case InsertionBehavior.ThrowOnExisting:
+                        ThrowHelper.ThrowDuplicateKey(key);
+                        break;
+
+                    default:
+                        return false;
```

Comment:
What's the insertion behavior handled by the default case? A Debug assert might better highlight this.

Reply:
TryAdd, where nothing happens if the key already exists.